firebase-firestore/src/main/java/com/google/firebase/firestore/util/Executors.java

```diff
@@ -29,6 +40,39 @@
   /** An executor that executes the provided runnable immediately on the current thread. */
   public static final Executor DIRECT_EXECUTOR = Runnable::run;
 
+  /**
+   * An executor that runs tasks in parallel on Android's AsyncTask.THREAD_POOL_EXECUTOR.
+   *
+   * <p>Unlike the main THREAD_POOL_EXECUTOR, this executor manages its own queue of tasks and can
+   * handle an unbounded number of pending tasks.
```

Comment:
Having an unbounded number of pending tasks seems problematic from a memory usage point of view.

Previously we'd unpack one blob, parse it, compare it, and then allow it to be GCed.

Now, if we get significantly ahead of the background threads' ability to clear this queue each queued document will be resident. Previously, ignoring constant factors around the current document we're looking at, our memory usage was essentially O(matching documents), but this increases it to potentially be O(all documents in a collection). This could be a huge difference. We should block admission into the queue if the background threads aren't cleaning up.

A straightforward way to bound this is to just use a semaphore with a fixed number of permits to guard admission. This will have the side effect of fixing the thread-safety issue you've documented below, which makes me uncomfortable.